array/array.go

```diff
@@ -0,0 +1,253 @@
+package array
+
+import (
+	"github.com/apache/arrow/go/arrow"
+	"github.com/apache/arrow/go/arrow/array"
+	"github.com/apache/arrow/go/arrow/memory"
+)
+
+//go:generate -command tmpl ../gotool.sh github.com/benbjohnson/tmpl
+//go:generate tmpl -data=@numeric.tmpldata -o numeric.gen.go numeric.gen.go.tmpl
+
+// Interface represents an immutable sequence of values.
+type Interface = array.Interface
+
+// Builder provides an interface to build arrow arrays.
+type Builder interface {
+	// Retain increases the reference count by 1.
+	// Retain may be called simultaneously from multiple goroutines.
+	Retain()
+
+	// Release decreases the reference count by 1.
+	Release()
+
+	// Len returns the number of elements in the array builder.
+	Len() int
+
+	// Cap returns the total number of elements that can be stored
+	// without allocating additional memory.
+	Cap() int
+
+	// NullN returns the number of null values in the array builder.
+	NullN() int
+
+	// AppendNull adds a new null value to the array being built.
+	AppendNull()
+
+	// Reserve ensures there is enough space for appending n elements
+	// by checking the capacity and calling Resize if necessary.
+	Reserve(n int)
+
+	// Resize adjusts the space allocated by b to n elements. If n is greater than b.Cap(),
+	// additional memory will be allocated. If n is smaller, the allocated memory may reduced.
+	Resize(n int)
+
+	// NewArray creates a new array from the memory buffers used
+	// by the builder and resets the Builder so it can be used to build
+	// a new array.
+	NewArray() Interface
+}
+
+type Boolean = array.Boolean
+
+type BooleanBuilder struct {
+	b *array.BooleanBuilder
+}
+
+func NewBooleanBuilder(mem memory.Allocator) *BooleanBuilder {
+	return &BooleanBuilder{
+		b: array.NewBooleanBuilder(mem),
+	}
+}
+func (b *BooleanBuilder) Retain() {
+	b.b.Retain()
+}
+func (b *BooleanBuilder) Release() {
+	b.b.Release()
+}
+func (b *BooleanBuilder) Len() int {
+	return b.b.Len()
+}
+func (b *BooleanBuilder) Cap() int {
+	return b.b.Cap()
+}
+func (b *BooleanBuilder) NullN() int {
+	return b.b.NullN()
+}
+func (b *BooleanBuilder) Append(v bool) {
+	b.b.Append(v)
+}
+func (b *BooleanBuilder) AppendValues(v []bool, valid []bool) {
+	b.b.AppendValues(v, valid)
+}
+func (b *BooleanBuilder) UnsafeAppend(v bool) {
+	b.b.UnsafeAppend(v)
+}
+func (b *BooleanBuilder) AppendNull() {
+	b.b.AppendNull()
+}
+func (b *BooleanBuilder) UnsafeAppendBoolToBitmap(isValid bool) {
+	b.b.UnsafeAppendBoolToBitmap(isValid)
+}
+func (b *BooleanBuilder) Reserve(n int) {
+	b.b.Reserve(n)
+}
+func (b *BooleanBuilder) Resize(n int) {
+	b.b.Resize(n)
+}
+func (b *BooleanBuilder) NewArray() Interface {
+	return b.NewBooleanArray()
+}
+func (b *BooleanBuilder) NewBooleanArray() *Boolean {
+	return b.b.NewBooleanArray()
+}
+
+type String struct {
+	value  string
+	length int
+	data   *array.Binary
+}
```

Comment:
Why don't you need an `isNull` field here?

Reply:
I generally thought it was unlikely to provide a lot of utility. In general, group keys aren't null very often if at all. If they do end up being null, the native arrow representation of them isn't very intense. It's a bitmap with the number of elements divided by 8 for the size.

I figured that adding an additional field to the struct would likely end up costing more in terms of memory overall because the situation where a string column is completely filled with null values would be so rare.

If we want to add `isNull` anyway I can modify this. I just figured we'd wait before optimizing that situation.

Reply:
Ok, fair enough. I agree it's not a common case. We don't just append to a bitmap though, we still need to reserve space for a pointer to the value that does not exist, right?